pkg/kubelet/volumemanager/reconciler/reconstruct_new.go

```diff
@@ -0,0 +1,243 @@
+// is populated.
+func (rc *reconciler) reconstructVolumes() {
+	defer rc.updateLastSyncTime()
+	// Get volumes information by reading the pod's directory
+	podVolumes, err := getVolumesFromPodDir(rc.kubeletPodsDir)
+	if err != nil {
+		klog.ErrorS(err, "Cannot get volumes from disk, skip sync states for volume reconstruction")
+		return
+	}
+	reconstructedVolumes := make(map[v1.UniqueVolumeName]*globalVolumeInfo)
+	reconstructedVolumeNames := []v1.UniqueVolumeName{}
+	for _, volume := range podVolumes {
+		if rc.actualStateOfWorld.VolumeExistsWithSpecName(volume.podName, volume.volumeSpecName) {
+			klog.V(4).InfoS("Volume exists in actual state, skip cleaning up mounts", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+			// There is nothing to reconstruct
+			continue
+		}
+		reconstructedVolume, err := rc.reconstructVolume(volume)
+		if err != nil {
+			klog.InfoS("Could not construct volume information", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName, "err", err)
+			// We can't reconstruct the volume. Remember to check DSW after it's fully populated and force unmount the volume when it's orphaned.
+			rc.volumesFailedReconstruction = append(rc.volumesFailedReconstruction, volume)
+			continue
+		}
+		klog.V(4).InfoS("Adding reconstructed volume to actual state and node status", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+		gvl := &globalVolumeInfo{
+			volumeName:        reconstructedVolume.volumeName,
+			volumeSpec:        reconstructedVolume.volumeSpec,
+			devicePath:        reconstructedVolume.devicePath,
+			deviceMounter:     reconstructedVolume.deviceMounter,
+			blockVolumeMapper: reconstructedVolume.blockVolumeMapper,
+			mounter:           reconstructedVolume.mounter,
+		}
+		if cachedInfo, ok := reconstructedVolumes[reconstructedVolume.volumeName]; ok {
+			gvl = cachedInfo
+		}
+		gvl.addPodVolume(reconstructedVolume)
+
+		reconstructedVolumeNames = append(reconstructedVolumeNames, reconstructedVolume.volumeName)
+		reconstructedVolumes[reconstructedVolume.volumeName] = gvl
+	}
+
+	if len(reconstructedVolumes) > 0 {
+		// Add the volumes to ASW
+		rc.updateStatesNew(reconstructedVolumes)
+
+		// The reconstructed volumes are mounted, hence a previous kubelet must have already put it into node.status.volumesInUse.
+		// Remember to update DSW with this information.
+		rc.volumesNeedReportedInUse = reconstructedVolumeNames
+		// Remember to update devicePath from node.status.volumesAttached
+		rc.volumesNeedDevicePath = reconstructedVolumeNames
+	}
+	klog.V(2).InfoS("Volume reconstruction finished")
+}
+
+func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeName]*globalVolumeInfo) {
+	for _, gvl := range reconstructedVolumes {
+		err := rc.actualStateOfWorld.MarkVolumeAsAttached(
+			//TODO: the devicePath might not be correct for some volume plugins: see issue #54108
+			gvl.volumeName, gvl.volumeSpec, "" /* nodeName */, gvl.devicePath)
```

Comment:
just to confirm nodeName is empty?

Reply:
This is copied from the old reconstruction: https://github.com/kubernetes/kubernetes/blob/3a79466ddde88e5161f1ca9754bebc62ef4ac8b9/pkg/kubelet/volumemanager/reconciler/reconstruct.go#L142-L143

`nodeName` in `MarkVolumeAsAttached` is then not used:
https://github.com/kubernetes/kubernetes/blob/e0f3e5c4575f2ad0c1a62d2b24c5ebe8483da5c0/pkg/kubelet/volumemanager/cache/actual_state_of_world.go#L361-L364

The parameter is there to satisfy `ActualStateOfWorldAttacherUpdater` [interface](https://github.com/kubernetes/kubernetes/blob/fc245b339b3a72307435f51c02e783eac7b66825/pkg/volume/util/operationexecutor/operation_executor.go#L240).

Now that I am looking at it, adding real node name is simple and would make the code more robust. Fixing in a new commit in both old/new reconstruction.